### PR TITLE
BatchAI. Support for 2018-03-01 API

### PR DESCRIPTION
**Pull request by AlexanderYukhanov**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by promptws** [comment]
View a preview at https://prompt.ws/r/Azure/azure-cli/5923
This is an **experimental preview** for @microsoft users.

File: src/command_modules/azure-cli-batchai/HISTORY.rst [hunk 1]
```diff
@@ -3,16 +3,37 @@
 Release History
 ===============
 
-0.1.6
-+++++
-* Support Autorest 3.0 based SDKs
-
-0.1.5
+1.0.0
```
Comment (tjprescott) [inline src/command_modules/azure-cli-batchai/HISTORY.rst]: Is this intended to be a GA release? If so, you'll need to schedule a full command review with the CLI team. Also... for reasons... the GA version is always "2.0.0".

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py [hunk 2]
```diff
@@ -31,31 +40,88 @@
...
 
+# Type of entries reported by list startup files.
+LogFile = collections.namedtuple('LogFile', 'name download_url is_directory size')
+
+logger = get_logger(__name__)
+
+
+def _get_resource_group_location(cli_ctx, resource_group):
+    client = get_mgmt_service_client(cli_ctx, ResourceType.MGMT_RESOURCE_RESOURCES)
+    return client.resource_groups.get(resource_group).location
```
Comment (tjprescott) [inline src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: You should use the "get_default_location_from_resource_group" that is included as part of the core.  This will throw a stack trace if the resource group does not exist.

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py [hunk 3]
```diff
@@ -31,31 +40,88 @@
...
+        return path
+    return None
+
 
 def _get_deserializer():
     client_models = {k: v for k, v in models.__dict__.items() if isinstance(v, type)}
     return Deserializer(client_models)
 
 
+def _ensure_resource_not_exist(client, resource_group, name):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: Why is this here? Create commands are expected to be idempotent. This would prevent that.

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py [hunk 4]
```diff
@@ -397,83 +693,163 @@ def list_jobs(client, resource_group=None):
...
-                       user_name=None, ssh_key=None, password=None, disk_count=None, disk_size=None,
-                       storage_sku='Premium_LRS', no_wait=False):
+def wait_for_job_completion(client, resource_group, job_name, check_interval_sec=15):
+    """Waits for specified job completion and setups the exit code to the code of the job.
+
+    :param azure.mgmt.batchai.BatchAIManagementClient client: Batch AI client.
+    :param str resource_group: name of the resource group.
+    :param str job_name: name of the job.
+    :param int check_interval_sec: how often to check the job state.
+    """
```
Comment (tjprescott) [inline src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: Help text should not added via docstrings on custom commands. These should be mirated to the `help.py` file or the `params.py` file

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py (hunk 3 above)
Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: Because we had a lot's of complains from the customers like "i submitting new job but it's not executing because i have already executed a job with such name o month ago" and asks to fix it

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py (hunk 4 above)
Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: we are not adding any help here. all command and parameters are documented in params.py

File: src/command_modules/azure-cli-batchai/HISTORY.rst (hunk 1 above)
Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/HISTORY.rst]: Can we schedule the review asap, please?

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py (hunk 2 above)
Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: thanks, will migrate

Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: looks like get_default_location_from_resource_group doesn't fit our needs, or i do not see how it fits

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py (hunk 4 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: Per our meeting, please just remove these docstrings.

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py (hunk 3 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: Let's let this in if it's helpful to customers. Part of my action items will be to have a discussion on our side to see if we can recommend a more helpful and consistent solution for these scenarios.

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py (hunk 2 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: Since you're using this method on multiple resource groups, then yes, the stock validator won't work and this is fine.

File: src/command_modules/azure-cli-batchai/HISTORY.rst (hunk 1 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-batchai/HISTORY.rst]: So as discussed we'll just call this release 0.1.6 or 0.2.0 while we work through a few more items that warrant further discussion within the CLI team.

File: src/command_modules/azure-cli-batchai/HISTORY.rst [hunk 5]
```diff
@@ -3,16 +3,40 @@
...
+* Added command line option to specify caching type for Batch AI managed NFS
+* Simplified specifying mount filesystem in config files. Now you can omit credentials for Azure File Share and
+  Azure Blob Containers - CLI will populate missing credentials using storage account key provided via command line
+  parameters or specified via environment variable or will query the key from Azure Storage (if the storage account
+  belongs to the current subscription).
+* Job file stream command now auto-completes when the job is completed (succeeded, failed, terminated or deleted).
+* Improved '-o table' support for show operations.
+* Added auto storage option for cluster creation.
+* Added ability to provide node setup task via command line.
+* 'job stream-file' and 'job list-files' commands are grouped under 'job file' group.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-batchai/HISTORY.rst]: Annotate this as a breaking change.

File: src/command_modules/azure-cli-batchai/HISTORY.rst [hunk 6]
```diff
@@ -3,16 +3,40 @@
...
+* Added more options to customize clusters - vm priority, subnet, initial nodes count for auto-scale clusters,
+  specifying custom image
+* Added command line option to specify caching type for Batch AI managed NFS
+* Simplified specifying mount filesystem in config files. Now you can omit credentials for Azure File Share and
+  Azure Blob Containers - CLI will populate missing credentials using storage account key provided via command line
+  parameters or specified via environment variable or will query the key from Azure Storage (if the storage account
+  belongs to the current subscription).
+* Job file stream command now auto-completes when the job is completed (succeeded, failed, terminated or deleted).
+* Improved '-o table' support for show operations.
+* Added auto storage option for cluster creation.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-batchai/HISTORY.rst]: We may have some more recommendations here prior to GA, but this is moving in the right direction and I don't see any reason to block this.

File: src/command_modules/azure-cli-batchai/HISTORY.rst (hunk 1 above)
Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/HISTORY.rst]: 0.2.0

File: src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py (hunk 4 above)
Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/azure/cli/command_modules/batchai/custom.py]: done

File: src/command_modules/azure-cli-batchai/HISTORY.rst (hunk 6 above)
Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/HISTORY.rst]: thanks

File: src/command_modules/azure-cli-batchai/HISTORY.rst (hunk 5 above)
Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/HISTORY.rst]: ok

File: src/command_modules/azure-cli-batchai/HISTORY.rst [hunk 7]
```diff
@@ -30,9 +30,14 @@ Release History
...
 * Job file stream command now auto-completes when the job is completed (succeeded, failed, terminated or deleted).
 * Improved '-o table' support for show operations.
-* Added auto storage option for cluster creation.
+* Added --use-auto-storage option for cluster creation. This option make it simpler to manage storage accounts and
+  and mount Azure File Share and Azure Blob Containers to clusters.
+* Added --generate-ssh-keys option into 'cluster create' and 'file-server create'.
 * Added ability to provide node setup task via command line.
-* 'job stream-file' and 'job list-files' commands are grouped under 'job file' group.
+* Braking change: 'job stream-file' and 'job list-files' commands are grouped under 'job file' group.
+* Braking change: renamed --admin-user-name to --user-name in 'file-server create' command to be consistent with
```
Comment (tjprescott) [inline src/command_modules/azure-cli-batchai/HISTORY.rst]: Typo:  Breaking

Comment (AlexanderYukhanov) [reply src/command_modules/azure-cli-batchai/HISTORY.rst]: thanks!
